### DISCUSSION TITLE
Shared database types in frontend and Deno edge functions

### DISCUSSION THREAD

**Discussion by Zizazorro** [supabase/supabase#28837] [Questions]
So I've made my first edge functions, but I want to use the database types that I use in my frontend in the edge functions as well. I followed the tutorial from the Supabase youtube channel and created a file (`frontend/src/app/global.d.ts`). It imports the supabase database types (which are generated automatically). My global.d.ts file looks like this:

```
import type { Database as DB } from "@/types/supabase";

declare global {
    type Database = DB;

    // Table row types
    type Game = DB["public"]["Tables"]["games"]["Row"];
}
```

This all works fine in my frontend (nextjs), but it doesn't seem to work in my edge functions, which are located in `frontend/supabase/functions`.

If I use the `Game` type in my edge functions, it says `type Game = DB`. It seems that it cannot correctly see the nested types from an import. If I copy paste my database types directly in the global.d.ts file, it **does** work. 

Things that I tried:

- Use import maps 
- Update the deno.json with `typeRoots` and `types` 
- Move the types folder to the top level of my project
- Directly export the type `Game` above the global declaration

Are there any examples where people use those shared types both in the nextjs frontend as in the Deno edge functions?

**Comment by shawnesquivel** [comment]
I'd love to know this as well. Part of using Edge Functions is to benefit from Typescript in frontend/backend, but without being easily able to import types, the benefits of using Edge Functions decreases significantly IMO. I've frequented the Edge Functions docs and it seems a tutorial in this manner would be highly loved by the community!

**Comment by davetorbeck** [comment]
It's quite puzzling there is no official documentation for this.  The DX for edge functions leaves a lot to be desired and I'm thinking of just moving to Cloudflare workers.

**Reply by shawnesquivel** [reply]
been using edge functions for the last 6mo. and 100% agree with you unfortunately. would love to contribute to this functionality if the Supabase team could help me get started!

**Comment by felixzieger** [comment]
Generate types in the `supabase/functions/_shared/` folder:
```
supabase gen types typescript --local > supabase/functions/_shared/supabase-types.ts
```

And use it in the edge function:
```typescript
import type { Database } from "../_shared/supabase-types.ts";

type SupabaseClient = ReturnType<typeof createClient<Database>>;
...
```

The frontend has it's own files, so I have a pnpm generate-types command that updates both files to not forget either.

Works fine for me.
